### PR TITLE
haskell-generic-builder: disable static PIE

### PR DESCRIPTION
**Pull request by rnhmjoj**
###### Motivation for this change

This fix the error
```
ld: -r and -pie may not be used together
```
when building static haskell packages using the `pkgsStatic` overlay.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via `pkgsStatic.haskellPackages.random`
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by sternenseemann — APPROVED** [review]

**Comment by rnhmjoj** [comment]
Moved to staging since it's a mass rebuild.

**Comment by sternenseemann** [comment]
We can do the mass rebuild on haskell-updates as well, that's no problem, probably better since it doesn't interfere with the staging workflow at all. Just up to @cdepillabout when he has time to wait for Hydra finishing 1-2 days.

File: pkgs/development/haskell-modules/generic-builder.nix
Comment (cdepillabout) [inline pkgs/development/haskell-modules/generic-builder.nix:425]: Can you turn `ghc/ghc#19580` into an actual link to the GHC issue tracker?

------------------------------------------------------

Also, I don't quite understand the combination of the comment here with the code.

The comment says: `Static libraries fails to build if Nix forcibly adds `-pie` to the linker flags`.  But then the code does appear to add `pie` to `hardeningDisable`?

**Comment by cdepillabout** [comment]
@rnhmjoj Could you show an example of using this PR to build a Haskell package statically?

File: pkgs/development/haskell-modules/generic-builder.nix
Comment (rnhmjoj) [reply pkgs/development/haskell-modules/generic-builder.nix:425]: It seems that by default Nix adds `-pie` to the linker flags, while GHC adds `-r -no-pie` which have the opposite meaning and so the linking fails. Adding `hardeningDisable = [ "pie" ]` prevent Nix from adding `-pie`: the option has a negative meaning.

**Comment by rnhmjoj** [comment]
> @rnhmjoj Could you show an example of using this PR to build a Haskell package statically?

Simply try `nix-build -A pkgsStatic.haskellPackages.random`: it should fail with the error I mentioned above.
I'm not sure what's causing the bug to appear, because not every haskell package is affected, but most are, though.

File: pkgs/development/haskell-modules/generic-builder.nix
Comment (rnhmjoj) [reply pkgs/development/haskell-modules/generic-builder.nix]: The comment should be a bit clearer, now.

**Comment by cdepillabout** [comment]
> `nix-build -A pkgsStatic.haskellPackages.random`

Ah, thanks, I didn't realize it was this easy.

I tried this on the current `haskell-updates` branch (at commit 00cca094c4d):

```console
$ nix-build -A pkgsStatic.haskellPackages.random
...
Preprocessing library for splitmix-0.1.0.3..
Building library for splitmix-0.1.0.3..
[1 of 4] Compiling Data.Bits.Compat ( src-compat/Data/Bits/Compat.hs, dist/build/Data/Bits/Compat.o )
[2 of 4] Compiling System.Random.SplitMix.Init ( src/System/Random/SplitMix/Init.hs, dist/build/System/Random/SplitMix/Init.o )
/nix/store/xwdw8df7hv4xgkvzjjkk1d26pg8fp8zd-x86_64-unknown-linux-musl-binutils-2.35.1/bin/x86_64-unknown-linux-musl-ld: -r and -pie may not be used together
`x86_64-unknown-linux-musl-ld' failed in phase `Merge objects'. (Exit code: 1)
builder for '/nix/store/38p7f7wcij9mkkcxm46wqgixw2smii4q-splitmix-0.1.0.3-x86_64-unknown-linux-musl.drv' failed with exit code 1
cannot build derivation '/nix/store/pckzibrv6lxbgvz37ryi99vvjxyi4dqm-random-1.2.0-x86_64-unknown-linux-musl.drv': 1 dependencies couldn't be built
error: build of '/nix/store/pckzibrv6lxbgvz37ryi99vvjxyi4dqm-random-1.2.0-x86_64-unknown-linux-musl.drv' failed
```

I appeared to get the same error you were seeing.

I then applied the patch from this PR, and re-ran the above Nix command:

(currently waiting for this to finish compiling.  I'll edit this when it hopefully successfully finishes)

**Comment by cdepillabout** [comment]
I cherry-picked this into https://github.com/NixOS/nixpkgs/pull/123682 as 830ef64.

You can watch the evaluation on Hydra if you're interested: https://hydra.nixos.org/eval/1671982.  Note that this also includes merging `master` into` haskell-updates`, so it is possible that build failure may be from that as well.  I imagine it is unlikely that any build failures would come from this static-linking fix.

**Comment by rnhmjoj** [comment]
Thank you!

**Comment by Mic92** [comment]
I am also looking into static-pie right now. It should be possible: https://github.com/NixOS/nixpkgs/pull/123989

**Comment by cdepillabout** [comment]
@rnhmjoj 

In the Haskell infrastructure, we have a file that defines a bunch of packages that are built by Hydra for the [`haskell-updates`](https://github.com/NixOS/nixpkgs/pull/123682) branch:

https://github.com/NixOS/nixpkgs/blob/d45bcb549ec7146148ea498a93178f48b9a1cfd6/pkgs/top-level/release-haskell.nix

If you look on Hydra you can see the builds for this jobset: https://hydra.nixos.org/jobset/nixpkgs/haskell-updates

We've been debating adding some sort of `pkgsStatic.haskellPackages.FOO` job to make sure that statically linking Haskell executables works at least somewhat.  Would you be willing to be added as a maintainer for a job like this?

If you're a maintainer, you'll be pinged on PRs like the above when the job breaks.  For example: https://github.com/NixOS/nixpkgs/pull/123682#issuecomment-845581510.  Here, you can see that the `taffybar` derivation is broken, and rvl has been pinged to fix it.

Is this something you'd be willing to do?

(Our guess is that static linking probably won't fail that often, so in practice this probably wouldn't require too much work.)

----------------------------------------------

Anyone else is of course welcome to volunteer to be a maintainer for this as well.  

cc @maralorn @sternenseemann @nh2

**Comment by rnhmjoj** [comment]
> I am also looking into static-pie right now. It should be possible: #123989

@Mic92 In general yes, but GHC can't produce this kind of executables, yet.

> We've been debating adding some sort of pkgsStatic.haskellPackages.FOO job to make sure that statically linking Haskell executables works at least somewhat. Would you be willing to be added as a maintainer for a job like this?

@cdepillabout I can't guarantee I'll be able to fix issues in a timely manner, but I'm trying to keep the static builds infrastructure working: I'm distributing static binaries for a couple of haskell programs I wrote, so whenever I make a new release I have to do this anyway. If this is ok for you then yes, you can add me as a maintainer.

@nh2 is also maintaining static builds, though I think their approach is to use `pkgsMusl` rather than `pkgsStatic` due to issues like TemplateHaskell.

**Comment by sternenseemann** [comment]
> @nh2 is also maintaining static builds, though I think their approach is to use pkgsMusl rather than pkgsStatic due to issues like TemplateHaskell.

How does that work `pkgsMusl` is very similar to `pkgsStatic` — except it doesn't link statically.

Also if musl is ever a problem, we have `glibc.static` by now, something to investigate as well at some point.

**Comment by rnhmjoj** [comment]
> How does that work pkgsMusl is very similar to pkgsStatic — except it doesn't link statically.

Correct, static linking is then enable manually. You can see an example here: https://github.com/nh2/static-haskell-nix/blob/master/default.nix

> Also if musl is ever a problem, we have glibc.static by now, something to investigate as well at some point.

Indeed, IIRC musl was chosen for `pkgsStatic` only because glibc was hard to statically link.

**Comment by nh2** [comment]
> @nh2 is also maintaining static builds, though I think their approach is to use `pkgsMusl` rather than `pkgsStatic` due to issues like TemplateHaskell.

That is correct, TH execution loads `.so` files by default so we need to have both `.so` and `.a` files around. `pkgsStatic` disables `.so` files where it can.

> Indeed, IIRC musl was chosen for `pkgsStatic` only because glibc was hard to statically link.

Yes, glibc maintainers do not really like / care about static linking, and it shows in some places, for example networking functionality.

**Comment by sternenseemann** [comment]
> Yes, glibc maintainers do not really like / care about static linking, and it shows in some places, for example networking functionality.

We have `glibc.static` in nixpkgs nowadays, although it does not seem to see much use. Changing `pkgsStatic` to use that by default could be an interesting experiment.

**Pull request closed without merging**
